File: src/player.zig
const Engine = @import("engine.zig");


const Self = @This();

sprite: Engine.Gfx.Sprite,
position: Engine.sdl.Point,

pub fn init(sprite: Engine.Gfx.Sprite, position: Engine.sdl.Point) Self {
    return Self {
        .sprite = sprite,
        .position = position,
    };
}

pub fn update(self: *Self, game_state: *Engine.GameState) !void {
    _ = self;
    _ = game_state;
    // game_state.gfx.camera_offset.x += 1;

}

pub fn draw(self: *Self, game_state: *Engine.GameState) !void {
    self.sprite.dst_rect.x = self.position.x;
    self.sprite.dst_rect.y = self.position.y;

    try game_state.gfx.draw_sprite(&self.sprite);
}

File: assets/scripts/modules/player.lua
local scancodes = require "assets.scripts.scancodes"

local Player = {}


function Player:start()
    self.position = point(2, 0)
    ---@class (exact) Sprite
    self.sprite = {}
    self.sprite.texture = AssetsManager.get_texture("entity")
    self.sprite.segment = rect(0, 0, 16, 16)
    self.sprite.dst_rect = rect(0, 0, 16, 16)
end

function Player:update()
    if Input.is_key_just_pressed(scancodes.SDL_SCANCODE_W) then
        self.position:move(0, -1)
    end
    if Input.is_key_just_pressed(scancodes.SDL_SCANCODE_S) then
        self.position:move(0, 1)
    end

    if Input.is_key_just_pressed(scancodes.SDL_SCANCODE_A) then
        self.position:move(-1, 0)
    end
    if Input.is_key_just_pressed(scancodes.SDL_SCANCODE_D) then
        self.position:move(1, 0)
    end

    Gfx.set_camera_offset(point(-self.position.x, -self.position.y))
end

function Player:draw()
    self.sprite.dst_rect:set_position(self.position.x, self.position.y)
    Gfx.draw_sprite(self.sprite)
end

return Player
